File: Kalorie/Kalorie/Core/Auth/AuthProvider.swift
//
//  AuthProvider.swift
//  Kalorie
//
//  Created by Josef Antoni on 30.06.2026.
//

import Foundation
import FirebaseAuth

enum AuthError: Error {
    case notAuthenticated
}

enum AuthProviderKind {
    case apple
    case google
}

protocol AuthProviderProtocol {
    var userId: String? { get }
    var isAnonymous: Bool { get }
    var displayName: String? { get }
    var lastSignInDate: Date? { get }
    var linkedProviderKind: AuthProviderKind? { get }
}

struct AuthProvider: AuthProviderProtocol {

    // MARK: - Properties

    var userId: String? {
        Auth.auth().currentUser?.uid
    }

    var isAnonymous: Bool {
        Auth.auth().currentUser?.isAnonymous ?? true
    }

    var displayName: String? {
        Auth.auth().currentUser?.displayName
    }

    var lastSignInDate: Date? {
        Auth.auth().currentUser?.metadata.lastSignInDate
    }

    var linkedProviderKind: AuthProviderKind? {
        // .first relies on an account never having more than one linked provider —
        // enforced by LinkOrMergeCredentialUseCase, not by Firebase Auth itself.
        switch Auth.auth().currentUser?.providerData.first?.providerID {
        case "apple.com": .apple
        case "google.com": .google
        default: nil
        }
    }
}

#if DEBUG
struct AuthProviderFake: AuthProviderProtocol {

    // MARK: - Properties

    var userId: String? = "test-user-id"
    var isAnonymous = true
    var displayName: String?
    var lastSignInDate: Date? = .now
    var linkedProviderKind: AuthProviderKind?
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/auth/AuthProvider.kt
package antoni.kalorie.core.auth

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import java.time.Instant

sealed class AuthError : Exception() {
    data object NotAuthenticated : AuthError()
}

enum class AuthProviderKind {
    GOOGLE,
}

interface AuthProviderProtocol {
    val userId: String?
    val isAnonymous: Boolean
    val displayName: String?
    val lastSignInDate: Instant?
    val linkedProviderKind: AuthProviderKind?
}

class AuthProvider(
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
) : AuthProviderProtocol {

    // MARK: - Properties

    override val userId: String?
        get() = auth.currentUser?.uid

    override val isAnonymous: Boolean
        get() = auth.currentUser?.isAnonymous ?: true

    override val displayName: String?
        get() = auth.currentUser?.displayName

    override val lastSignInDate: Instant?
        get() = auth.currentUser?.metadata?.lastSignInTimestamp?.let(Instant::ofEpochMilli)

    override val linkedProviderKind: AuthProviderKind?
        get() = when (auth.currentUser?.providerData?.firstOrNull { it.providerId != FIREBASE_PROVIDER_ID }?.providerId) {
            GoogleAuthProvider.PROVIDER_ID -> AuthProviderKind.GOOGLE
            else -> null
        }

    private companion object {
        const val FIREBASE_PROVIDER_ID = "firebase"
    }
}
